### DISCUSSION TITLE
Need help resolving: "missing module named System - (serial)" and more.

### DISCUSSION THREAD

**Discussion by LizThompson93** [pyinstaller/pyinstaller#7327] [Help] [answered]
Hello, I have been pouring over Google, Reddit, and StackOverflow but I just can't seem to find the solution to my problem. I am using Windows 10, Anaconda, and am having problem with pySerial:

`import serial`

I am getting the warning statements for the following:

> missing module named termios - imported by getpass (optional), serial.serialposix (top-level)
> missing module named fcntl - imported by serial.serialposix (top-level)
> missing module named 'System.IO' - imported by serial.serialcli (top-level)
> missing module named System - imported by serial.serialcli (top-level)

I've even tried pointing directly to the serial folder with my .spec file:

```
a = Analysis(
    ['My_Program.py'],
    pathex=["anaconda3/Lib/site-packages/serial"],
    binaries=[],
    datas=[],
    hiddenimports=[],
    hookspath=['serial\\\\__init__.py'],
    hooksconfig={},
    runtime_hooks=[],
    excludes=[],
    win_no_prefer_redirects=False,
    win_private_assemblies=False,
    cipher=block_cipher,
    noarchive=False,
)
```
Does anyone have advice on resolving this issue?

**Comment by rokm** [comment, marked as answer]
Those are just warnings, because PyInstaller found (conditional) imports (or rather, imports in `pyserial` backend modules that are used conditionally) but cannot resolve them.

 `termios` and `fcntl` are POSIX specific, so they are not available on Windows (they are imported from `serialposix` module is which is not used on Windows). 

`System` and `System.IO` are imported in the .NET based backend, which you are also unlikely to be using (otherwise you would have those modules available).

So these warnings should not affect your program in any way. If you're having problems with it, the reason is elsewhere.

---

That said...

>     pathex=["anaconda3/Lib/site-packages/serial"],

This is not how search paths work in python. If you need to extend search path, it should be the directory containing the packages (i.e., path to `site-packages` directory), not the package directory itself. And you shouldn't need to add standard environment's package directories to search paths, unless you're doing something shifty, like mixing up different python environments (running pyinstaller that's installed in different environment than the rest of your packages).

Adding a package directory to the search path is problematic, because it can cause the package's sub-modules to become top-level modules, and confuse the import analysis.

>     hookspath=['serial\\\\__init__.py'],

This argument is used to pass paths to directories containing extra PyInstaller hooks. So passing it a path to package's `__init__.py` does nothing.

**Reply by LizThompson93** [reply]
Hello Rokm, thank you for your response and clarification on modifying .spec files! So I converted my Jupyter .ipynb to .py using nbconvert and then turning it into a .exe with pyinstaller. I have deleted my .spec file and am just using the default command now:

`pyinstaller "MyPythonProgram.py"`

The error I see when trying to run is:
```
Error loading Python DLL 'C:\Users\Liz\build\MyPythonProgram\python38.dll'.
LoadLibrary: The specified module could not be found.
```

I went ahead and copy and pasted python38.dll from my anaconda3 folder and plopped it in and then tried running it again:

```
Traceback (most recent call last):
  File "MyPythonProgram.py", line 8, in <module>
  File "PyInstaller\loader\pyimod02_importers.py", line 499, in exec_module
  File "serial\__init__.py", line 29, in <module>
  File "PyInstaller\loader\pyimod02_importers.py", line 499, in exec_module
  File "serial\serialwin32.py", line 15, in <module>
  File "PyInstaller\loader\pyimod02_importers.py", line 499, in exec_module
  File "ctypes\__init__.py", line 7, in <module>
ModuleNotFoundError: No module named '_ctypes'
[17584] Failed to execute script 'MyPythonProgram' due to unhandled exception!

```

I only have 4 import statements:
```
import time #Line 7
import serial #Line 8
from ctypes import windll #Line 9
from win32gui import GetWindowText, GetForegroundWindow #Line 10
```

I have ensured I only have one python.exe which is from my Anaconda3 folder. These errors led me to believe that it was a problem with pySerial's import serial. What are your thoughts?

**Reply by LizThompson93** [reply]
Hm, I googled and looks like PyInstaller and ctypes don't play nice together. I see here a snippet:

https://pyinstaller.org/en/stable/feature-notes.html
"...it will search for simple standard usages of ctypes and automatically track and bundle the referenced libraries. "

I'm a little bit confused as to what I add to

`binaries=[],`

In my .spec files?

**Reply by bwoodsend** [reply]
You're on the wrong track. Those docs are talking about libraries loaded using ctypes whereas your problem is that ctypes itself (or more likely its dependency `libcffi.so`) is missing. That library should be somewhere in `anaconda3/Library/bin`. This is such a Conda issue though - you'd have a much easier time using normal Python from https://python.org/downloads.

**Reply by rokm** [reply]
> Error loading Python DLL 'C:\Users\Liz\build\MyPythonProgram\python38.dll'.

The final executable and all accompanying files are in the `dist` directory. `build` contains intermediate results, so trying to run an executable in there will result in missing libraries.

**Reply by LizThompson93** [reply]
> > Error loading Python DLL 'C:\Users\Liz\build\MyPythonProgram\python38.dll'.
> 
> The final executable and all accompanying files are in the `dist` directory. `build` contains intermediate results, so trying to run an executable in there will result in missing libraries.

Oh god, that was the problem. I was just looking at the terminal outputs and just followed where it was writing which was "build" when I should have been using "dist" files. I feel embarrassed but thank you very much Rokm, Bwoodsend!
